File: uniffi/src/bindings/swift/templates/EnumTemplate.swift
public enum {{ e.name() }}: Lowerable, Liftable {
    {% for value in e.values() %}
    case {{ value|decl_enum_variant_swift }}
    {% endfor %}

    static func lift(from buf: Reader) throws -> {{ e.name() }} {
        return try {{ e.name() }}.fromFFIValue(UInt32.lift(from: buf))
    }

    static func fromFFIValue(_ number: UInt32) throws -> {{ e.name() }} {
        switch number {
        {% for value in e.values() %}
        case {{ loop.index }}: return .{{ value|decl_enum_variant_swift }}
        {% endfor %}
        default: throw InternalError.unexpectedEnumCase
        }
    }

    func lower(into buf: Writer) {
        self.toFFIValue().lower(into: buf)
    }

    func toFFIValue() -> UInt32 {
        switch self {
        {% for value in e.values() %}
        case .{{ value|decl_enum_variant_swift }}: return {{ loop.index }}
        {% endfor %}
        }
    }
}

File: uniffi/src/bindings/kotlin/templates/EnumTemplate.kt
enum class {{ e.name()|class_name_kt }} {
    {% for value in e.values() %}
    {{ value|enum_name_kt }}{% if loop.last %};{% else %},{% endif %}
    {% endfor %}

    companion object {
        internal fun lift(n: Int) =
            try { values()[n - 1] }
            catch (e: IndexOutOfBoundsException) {
                throw RuntimeException("invalid enum value, something is very wrong!!", e)
            }

        internal fun liftFrom(buf: ByteBuffer) = lift(Int.liftFrom(buf))
    }

    internal fun lower() = this.ordinal + 1

    internal fun lowersIntoSize() = 4

    internal fun lowerInto(buf: ByteBuffer) = this.lower().lowerInto(buf)
}